autorally_core/src/StateEstimator/StateEstimator.cpp

```diff
@@ -284,42 +284,113 @@ namespace autorally_core
     gyro = Vector3(gx, gy, gz);
   }
 
-  void Imu_Gps::GpsHelper()
+  void StateEstimator::GpsHelper()
   {
+    double prevTime = 0;
+    double curTime = 0;
+    double timeWithoutGPS = 0;
+    unsigned char status = autorally_msgs::stateEstimatorStatus::OK;
+    ros::Rate loop_rate(m_frequency * 1.25);
+    int gpsErrorCount = 0;
+
     // Kick off the thread, and wait for our GPS measurements to come streaming in
     while (ros::ok())
     {
-      sensor_msgs::NavSatFixConstPtr fix = m_gpsOptQ.popBlocking();
-      ++m_gpsCounter;
-      if (m_gpsCounter > m_gpsSkip) m_gpsCounter = 0;
-      else continue;
+      sensor_msgs::NavSatFixConstPtr fix;
+      bool usingGPS = false;
+      bool usingOdom = false;
+      bool usingIMU = false;
+
+
+      // remove any old GPS measurements
+      while (m_gpsOptQ.size() > 0 && m_gpsOptQ.front()->header.stamp.toSec() < prevTime)
+        m_gpsOptQ.popBlocking();
+
+      // remove any old odom messages
+      while (m_usingOdom && m_odomOptQ.size() > 0 && m_odomOptQ.front()->header.stamp.toSec() < prevTime)
+        m_lastOdom = m_odomOptQ.popBlocking();
+
+
+      // set up for using odom or GPS timestamp but always at ~10Hz
+      if (m_gotFirstFix && m_gpsOptQ.size() > 0 && (m_gpsOptQ.back()->header.stamp.toSec() - prevTime) > 0.9/m_frequency)
+      {
+        fix = m_gpsOptQ.popBlocking();
+        while ((fix->header.stamp.toSec() - prevTime) < 0.9/m_frequency)
+          fix = m_gpsOptQ.popBlocking();
+
+        // check if the GPS measurement is close to expected from the previous velocity and pose
+        double E, N, U, expectedE, expectedN, expectedU, distance;
+        m_enu.Forward(fix->latitude, fix->longitude, fix->altitude, E, N, U);
+        expectedE = (curTime-prevTime)*m_prevVel[0] + m_prevPose.x();
+        expectedN = (curTime-prevTime)*m_prevVel[1] + m_prevPose.y();
+        expectedU = (curTime-prevTime)*m_prevVel[2] + m_prevPose.z();
+        distance = sqrt(pow(expectedE-E,2) + pow(expectedN-N,2) + pow(expectedU-U,2));
+        // if there has been no GPS then our estimate is inaccurate and should trust the GPS more
+        if ((distance < m_maxGPSError) || (timeWithoutGPS > m_timeWithoutGPS))
+        {
+          usingGPS =  true;
+          timeWithoutGPS = 0;
+          curTime = fix->header.stamp.toSec();
+        }
+        else
+        {
+          ROS_WARN("GPS message does not match expected position");
+          diag("GPS error count", std::to_string(++gpsErrorCount));
+        }
+      }
+      else if (m_usingOdom && m_gotFirstFix && (m_odomOptQ.size() > 0)
+          && (m_odomOptQ.back()->header.stamp.toSec() - prevTime) > 1.25/m_frequency)
+      {
+        // GPS messages are a little slower so wait a little extra before continuing without one
```

Comment:
Also, how would we account for the GPS timestamp being different than the timestamps used for the IMU (possibly with odom) factor? Without the GPS measurement there already, we wouldn't know what its timestamp will be. Therefore, we wouldn't necessarily have a position variable to add the GPS measurement to.